### PR TITLE
Add mempool

### PR DESCRIPTION
**Pull request by erikarvstedt**
This PR tracks my modifications to #456 (nixbitcoin:mempool-nb).
It is based on https://github.com/fort-nix/nix-bitcoin/pull/504 (`Add fulcrum`).
Deployed at https://mempool.nixbitcoin.org

### CODE REVIEW

**Comment by nixbitcoin** [comment]
I think this pretty much works now and could be spun out into a new PR to merge.

File: modules/mempool.nix [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+    # Available options:
+    # https://github.com/mempool/mempool/blob/master/backend/src/config.ts
+    services.mempool.settings = {
+      MEMPOOL = {
+        # mempool doesn't support regtest
+        NETWORK = "mainnet";
+        BACKEND = "electrum";
+        HTTP_PORT = cfg.backendPort;
+        CACHE_DIR = "/run/mempool";
```
Comment (nixbitcoin) [inline modules/mempool.nix:122]: This should be a permanent directory so cache doesn't get wiped on restart.

File: modules/presets/enable-tor.nix [hunk 2]
```diff
@@ -28,12 +28,17 @@ in {
     # btcpayserver = defaultEnableTorProxy;
     spark-wallet = defaultEnableTorProxy;
     lightning-pool = defaultEnableTorProxy;
+    # TODO-EXTERNAL:
+    # disable Tor enforcement until
+    # https://github.com/mempool/mempool/pull/1213 is released
```
Comment (nixbitcoin) [inline modules/presets/enable-tor.nix:33]: We can fetch rates over Tor now

File: modules/netns-isolation.nix [hunk 3]
```diff
@@ -293,6 +293,17 @@ in {
       clightning-rest = {
         id = 30;
       };
+      fulcrum = {
+        id = 31;
+        connections = [ "bitcoind" ];
+      };
+      mysql = {
```
Comment (nixbitcoin) [inline modules/netns-isolation.nix:300]: mysql doesn't need netns if we're using socket

File: modules/netns-isolation.nix [hunk 4]
```diff
@@ -293,6 +293,17 @@ in {
...
+        id = 32;
+      };
+      mempool = {
+        id = 33;
+        connections = [ "bitcoind" "electrs" "fulcrum" "nginx" "mysql" ];
```
Comment (nixbitcoin) [inline modules/netns-isolation.nix:305]: Connections to `mysql` netns don't need to be allowed since we're now using a UNIX socket

**Comment by erikarvstedt** [comment]
This PR is ready for review. 
Here's the [unsquashed version](https://github.com/erikarvstedt/nix-bitcoin/commits/mempool-ea-unsquashed).

Upstream requirements for merging:
- https://github.com/mempool/mempool/pull/2096
- https://github.com/mempool/mempool/pull/2078
- https://github.com/mempool/mempool/pull/2058

**Comment by wiz** [comment]
Thanks for working on this. Will get back to on those PRs soon.

**Comment by erikarvstedt** [comment]
To explicitly check the change to nodeinfo (which is covered by the tests), run:
```bash
run-tests.sh -s '{
  imports = [ scenarios.regtestBase ];
  services.mempool.enable = true;
  nix-bitcoin.onionServices.mempool-frontend.enable = true;
  nix-bitcoin.nodeinfo.enable = true;
}' container --run c nodeinfo

# Result:
# {
#   ...
#   "mempool-frontend": {
#     "local_address": "127.0.0.1:60845",
#     "onion_address": "udiacmki6rjrtcxkttwpv3iilczaz52gybodi7ceqwmubulhndhrapad.onion:80"
#   }
# }
```

File: examples/configuration.nix [hunk 5]
```diff
@@ -150,6 +161,28 @@
   # public access to the web interface.
   # nix-bitcoin.onionServices.btcpayserver.enable = true;
 
+  ### MEMPOOL
+  # Set this to enable mempool, a fully featured Bitcoin visualizer, explorer,
+  # and API service.
+  #
+  # services.fulcrum.enable = true;
```
Comment (nixbitcoin) [inline examples/configuration.nix:168]: Should this be `services.mempool.enable = true;`?

File: examples/configuration.nix [hunk 6]
```diff
@@ -150,6 +161,28 @@
...
+  # and API service.
+  #
+  # services.mempool.enable = true;
+  #
+  # Possible options for the Electrum backend server:
+  #
+  # - electrs:
+  #   Small database size, slow when querying new addresses.
+  #   Enable with:
+  #     services.mempool.electrumServer = "electrs";
```
Comment (nixbitcoin) [inline examples/configuration.nix:175]: Isn't this already enabled by default? Why list it here explicitly? This gives the user the idea that he has to explicitly enable it.

**Comment by nixbitcoin** [comment]
After reviewing the code and testing the module, I think this PR is ready to get merged. Let's wait for the upstream PRs and then merge this. :tada:

**Comment by wiz** [comment]
Sorry for the delay on merging the upstream PRs but they are in the queue

**Comment by wiz** [comment]
I think we are still waiting to merge a big nginx PR ?

**Comment by erikarvstedt** [comment]
> I think we are still waiting to merge a big nginx PR ?

No, this PR is only blocked by https://github.com/mempool/mempool/issues/2246, which is probably an upstream bug.
You recently downgraded the `axios` dependency (the possible cause of the bug), so let's give this another try.

**Comment by wiz** [comment]
@knorrium have you observed the axios state bug since we downgraded the dependency version?

**Comment by prusnak** [comment]
Fixes https://github.com/fort-nix/nix-bitcoin/issues/394

**Comment by prusnak** [comment]
Closes https://github.com/fort-nix/nix-bitcoin/issues/394

**Comment by erikarvstedt** [comment]
The bug has reappared, after running for 26 days.
I've restarted the backend.

What we'll do now:
- Release mempool as an extension flake in a separate repo, with a weekly timer that restarts the backend.
- Add an extra instance to nixbitcoin.org that can be used for debugging.

**Comment by wiz** [comment]
hmm that suggests the old bug we had where axios would get into a broken state is still around, easier to reproduce using tor - @knorrium after the umbrel release we should prioritize investigating this

**Comment by erikarvstedt** [comment]
The [mempool flake](https://github.com/erikarvstedt/nix-bitcoin-mempool) will be released when https://github.com/fort-nix/nix-bitcoin/pull/586 is merged.

I've added two extra mempool instances to nixbitcoin.org.
- https://mempool-debug-tor.nixbitcoin.org
  This sets a Tor `SOCKS5PROXY`, like in our main config.
- https://mempool-debug.nixbitcoin.org
  This runs without a Tor proxy. We've never run this in production over an extended period of time.
  Maybe the bug doesn't happen here.

They share the same bitcoind and fulcrum backend, and are [implemented here](https://github.com/fort-nix/nixbitcoin.org/commits/mempool-debug).
  
@wiz, when the bug reappears, we'll keep these instances running and give you SSH access to the nodejs debug socket.

**Comment by erikarvstedt** [comment]
The [mempool extension module](https://github.com/fort-nix/nix-bitcoin-mempool) has been released for some time, and is now listed in the [nix-bitcoin README](https://github.com/fort-nix/nix-bitcoin#extension-modules).

**Comment by wiz** [comment]
Thanks for the hard work on this. We've also made progress tracking down the remaining issues, it seems we weren't using axios in all places and the built-in NodeJS http client is buggy 😓

**Comment by erikarvstedt** [comment]
The bug has now appeared on both debug instances:
https://mempool-debug.nixbitcoin.org/address/1CGG9qVq2P6F7fo6sZExvNq99Jv2GDpaLE
https://mempool-debug-tor.nixbitcoin.org/address/1CGG9qVq2P6F7fo6sZExvNq99Jv2GDpaLE

Both instances were started on 2023-02-24 and failed ~20 days (debug-tor) and ~25 days (debug) later.
@wiz, @knorrium, if you're interested in nodejs debug socket access via SSH, let me know via Matrix.

**Comment by wiz** [comment]
Can you upgrade to v2.5.0 tag? We added a potential fix for this issue

**Comment by erikarvstedt** [comment]
I've updated the [extension module](https://github.com/fort-nix/nix-bitcoin-mempool) to the latest release `2.5.0`.

Also, I've updated both debug instances. If they run stably for 30 days, I'll add mempool as a core nix-bitcoin module.

@jb55, the extension module is ready for production use.

**Comment by jb55** [comment]
@erikarvstedt not gonna lie I have no idea how to extend my nixos system modules, I've always just patched my nixpkgs.

**Comment by jb55** [comment]
ah cool there are instructions. thank you!

**Comment by erikarvstedt** [comment]
The Electrum backend failure bug [has been fixed](https://github.com/mempool/mempool/issues/2246). Ready to merge.
This branch is running live on nixbitcoin.org

Use this for comparing the content of this PR with the extension flake:
```bash
git clone https://github.com/fort-nix/nix-bitcoin-mempool /tmp/nix-bitcoin-mempool
extension=/tmp/nix-bitcoin-mempool
nb=<nix-bitcoin-src>
git diff --no-index {$extension,$nb}/modules/mempool.nix
git diff --no-index {$extension,$nb}/pkgs/mempool
```

**Comment by erikarvstedt** [comment]
Rebased.

**Comment by chrisguida** [comment]
Hmm is this active on mempool.nixos.org rn? Address queries appear to be failing.

Screenshot from 2023-09-06 10-35-43

File: modules/mempool.nix [hunk 7]
```diff
@@ -0,0 +1,331 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  options.services = {
+    mempool = {
+      enable = mkOption {
```
Comment (seberm) [inline modules/mempool.nix:7]: Is there a reason to not use a `mkEnableOption` here?

Comment (erikarvstedt) [reply modules/mempool.nix:7]: Mempool has an extended, multi-sentence description text which doesn't fit the description template of `mkEnableOption`.

File: pkgs/mempool/generate.sh [hunk 8]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env nix-shell
+#! nix-shell -i bash -p gnupg gnused jq
+set -euo pipefail
```
Comment (jonasnick) [inline pkgs/mempool/generate.sh:3]: Is it expected that the hashes in "frontend-assets-fetch.sh" and `frontendAssets` in "mempool/default.sh" change when running "generate.sh"? It seems like for the frontend assets we look for there most recent ones regardless of the mempool version we try to "generate".

File: README.md [hunk 9]
```diff
@@ -103,12 +104,6 @@ NixOS modules ([src](modules/modules.nix))
   * [backups](modules/backups.nix): duplicity backups of all your node's important files
   * [operator](modules/operator.nix): configures a non-root user who has access to client tools (e.g. `bitcoin-cli`, `lightning-cli`)
 
-### Extension modules
-Extension modules are maintained in separate repositories and have their own review
-and release process.
-
-* [Mempool](https://github.com/fort-nix/nix-bitcoin-mempool): Bitcoin visualizer, explorer and API service
-
```
Comment (jonasnick) [inline README.md:111]: Do you think it makes sense to keep the extension module section and maybe link to an archived version of the mempool module?

File: modules/mempool.nix [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+        internal = true;
+        readOnly = true;
+        inherit default;
+      };
+    in {
+      enable = copyValue cfg.frontend.enable;
+      address = copyValue cfg.frontend.address;
+      port = copyValue cfg.frontend.port;
+    };
+  };
```
Comment (jonasnick) [inline modules/mempool.nix:151]: I don't quite understand how this works. onion-services.nix sets mempool-frontend `externalPort` and the `port` is the default `mempool.frontend.port`. What if the user sets `mempool.frontend.port`? In that cas, doesn't that lead to the creating an onion service for the wrong port?

Comment (erikarvstedt) [reply modules/mempool.nix]: No, `mempool.frontend.port` always equals `mempool-frontend.port`. I've renamed `copyValue` to `mkAlias`.

File: README.md (hunk 9 above)
Comment (erikarvstedt) [reply README.md:111]: No. What would be the benefit of this?

Comment (jonasnick) [reply README.md:111]: To document that extension modules are possible and how to create them.

File: pkgs/mempool/generate.sh (hunk 8 above)
Comment (erikarvstedt) [reply pkgs/mempool/generate.sh:3]: The frontend assets are not linked to a specific mempool version and are updated continuously.
I've removed the liquid assets. These are unused (we don't support liquid) and are updated daily.
Now we're left with mining pool assets, which only change every few months.

**Comment by erikarvstedt** [comment]
Rebased and deployed.

**Comment by erikarvstedt** [comment]
Fixed.

File: README.md (hunk 9 above)
Comment (erikarvstedt) [reply README.md:111]: I've added `nix-bitcoin-mempool` to the examples.

**Review by jonasnick — APPROVED** [review]
ACK 1de259485b76d787c5876829fc08da9064e03546

**Comment by jb55** [comment]
On Sun, Nov 05, 2023 at 01:41:24AM -0800, Jonas Nick wrote:
>Merged #505 into master.

🎉

**Comment by Fleinsopp** [comment]
File change diff

**Pull request merged**